peachjam/models/core_document_model.py

```diff
@@ -756,17 +756,37 @@ class SourceFile(AttachmentAbstractModel):
         _("source URL"), max_length=2048, null=True, blank=True
     )
 
+    file_as_pdf = models.FileField(
+        _("file as pdf"),
+        upload_to=file_location,
+        max_length=1024,
+        null=True,
+        blank=True,
+    )
+
     class Meta:
         verbose_name = _("source file")
         verbose_name_plural = _("source files")
 
     def as_pdf(self):
         if self.filename.endswith(".pdf"):
             return self.file
+        elif self.file_as_pdf:
+            return self.file_as_pdf
+        else:
+            return None
 
-        # convert with soffice
+    def convert_to_pdf(self):
         suffix = os.path.splitext(self.filename)[1].replace(".", "")
```

Comment:
i think this should also check if it has to do this work.

Based on the spec:

> The background task should first check if the source file already has a PDF version, and do nothing if it does (to make the task idempotent).

It's also a good approach for async tasks: the point of this method isn't to blindly create the pdf, it's to ensure that a PDF exists. If, for some reason, it already exists, it can stop.

Reply:
Makes sense. Added.